src/frontend/Canonicalize.ml

```diff
@@ -0,0 +1,243 @@
+let rec no_parens {expr; emeta} =
+  match expr with
+  | Paren e -> no_parens e
+  | Variable _ | IntNumeral _ | RealNumeral _ | GetLP | GetTarget ->
+      {expr; emeta}
+  | TernaryIf _ | BinOp _ | PrefixOp _ | PostfixOp _ ->
+      {expr= map_expression keep_parens ident expr; emeta}
+  | Indexed (e, l) ->
+      { expr=
+          Indexed
+            ( keep_parens e
+            , List.map
+                ~f:(function
+                  | Single e -> Single (no_parens e)
+                  | i -> map_index keep_parens i)
+                l )
+      ; emeta }
+  | ArrayExpr _ | RowVectorExpr _ | FunApp _ | CondDistApp _ ->
+      {expr= map_expression no_parens ident expr; emeta}
+
+and keep_parens {expr; emeta} =
+  match expr with
+  | Paren {expr= Paren e; _} -> keep_parens e
+  | Paren ({expr= BinOp _; _} as e)
+   |Paren ({expr= PrefixOp _; _} as e)
+   |Paren ({expr= PostfixOp _; _} as e)
+   |Paren ({expr= TernaryIf _; _} as e) ->
+      {expr= Paren (no_parens e); emeta}
+  | _ -> no_parens {expr; emeta}
+
+let parens_lval = map_lval_with no_parens ident
+
+let rec parens_stmt {stmt; smeta} =
+  let stmt =
+    match stmt with
+    | VarDecl
+        { decl_type= d
+        ; transformation= t
+        ; identifier
+        ; initial_value= init
+        ; is_global } ->
+        VarDecl
+          { decl_type= Middle.Type.map no_parens d
+          ; transformation= Middle.Program.map_transformation keep_parens t
+          ; identifier
+          ; initial_value= Option.map ~f:no_parens init
+          ; is_global }
+    | For {loop_variable; lower_bound; upper_bound; loop_body} ->
+        For
+          { loop_variable
+          ; lower_bound= keep_parens lower_bound
+          ; upper_bound= keep_parens upper_bound
+          ; loop_body= parens_stmt loop_body }
+    | _ -> map_statement no_parens parens_stmt parens_lval ident stmt
+  in
+  {stmt; smeta}
+
+let repair_syntax program : untyped_program =
+  map_program
+    (repair_syntax_stmt (userdef_distributions program.functionblock))
+    program
+
+let canonicalize_program program : typed_program =
+  String.Table.clear deprecated_userdefined ;
+  program.functionblock
+  |> Option.iter
+       ~f:
+         (List.iter ~f:(function
+           | { stmt=
+                 FunDef {funname= {name; _}; arguments= (_, type_, _) :: _; _}
+             ; smeta= _ }
+             when String.is_suffix ~suffix:"_log" name ->
+               let newname =
+                 if String.is_suffix ~suffix:"_cdf_log" name then
+                   String.drop_suffix name 8 ^ "_lcdf"
+                 else if String.is_suffix ~suffix:"_ccdf_log" name then
+                   String.drop_suffix name 9 ^ "_lccdf"
+                 else if Middle.UnsizedType.is_real_type type_ then
+                   String.drop_suffix name 4 ^ "_lpdf"
+                 else String.drop_suffix name 4 ^ "_lpmf"
+               in
+               String.Table.add deprecated_userdefined ~key:name ~data:newname
```

Comment:
Do we actually want to ignore duplicates or do we want to do something with them? I'll merge for now but just food for thought going forwards.

Reply:
Duplicates come from forward declarations (which is needed if the function is recursive). The signatures given in function declaration and definition must be identical and no other function may have the same name.